### PR TITLE
Libimaglink/init

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Tracking branch for implementing `libimaglink` which was proposed via #167 

This library implements _almost_ everything what would be needed to implement a "bookmarking" module, though it does focus on the _linking_ part and does not allow tagging and such.

The functionality of this library will be useful in other libraries and modules.

---

```
- [ ] `imag::/foo/bar~0.2.0` [requirement](https://github.com/matthiasbeyer/imag/blob/master/doc/src/02000-store.md#store-path-links-secthestorelinks)
```

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
@TheNeikos , @neithernut is such a functionality wanted? I think so, right?

File: libimaglink/src/internal.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            let links : Vec<Link> = links.into();
+            let links = links.into_iter().map(|link| Value::String(link.into())).collect();
+
+            let old_links = content_section.insert(String::from("links"), Value::Array(links));
+            if old_links.is_none() {
+                return Ok(Links::new(vec![]));
+            }
+            let old_links = old_links.unwrap();
+
+            if match old_links { Value::Array(_) => { false }, _ => { true } } {
```
Comment (matthiasbeyer) [inline libimaglink/src/internal.rs]: @TheNeikos as this scheme gets more frequent: Would it be possible to implement a `is_type!()` macro which does basically this?

Comment (TheNeikos) [inline libimaglink/src/internal.rs]: I would do it like this:

```rust
if let Value::Array(_) = old_links { false } else { true } 
```

Comment (matthiasbeyer) [inline libimaglink/src/internal.rs]: Ah, sure!

Anyways, how about an macro for this?

Comment (TheNeikos) [inline libimaglink/src/internal.rs]: That is not that easily possible really.

**Comment by matthiasbeyer** [comment]
Note: Merging this one after #149 will break the build, as we need the `store::EntryHeader::toml()` function here, which is removed in #149 .

**Comment by matthiasbeyer** [comment]
This is on hold until #149 is merged, as this PR would greatly simplify this PR.

**Comment by matthiasbeyer** [comment]
waiting on #194

**Comment by matthiasbeyer** [comment]
As nobody seems to care in #194, I implemented two-way linking for internal links.

I _hope_ that someone reviews this ASAP.

**Comment by neithernut** [comment]
It looks good on the first shallow read and some of the functionality introduced _is_ needed.

File: libimaglink/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            cause: cause,
+        }
+    }
+
+}
+
+impl Display for LinkError {
+
+    fn fmt(&self, fmt: &mut Formatter) -> Result<(), FmtError> {
+        try!(write!(fmt, "[{}]", link_error_type_as_str(&self.kind.clone())));
```
Comment (mario-kr) [inline libimaglink/src/error.rs]: is the clone necessary? argument is passed as reference.

Comment (matthiasbeyer) [inline libimaglink/src/error.rs]: Nope, thanks for spotting this one (the other one was fixed as well)

**Comment by matthiasbeyer** [comment]
I just rebased and cleaned up the history. I would keep the commits which are there now, even if there was a rewrite in between.

File: libimaglink/src/external.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+}
+
+pub trait ExternalLinker {
+
+    /// get the external link from the implementor object
+    fn get_external_link(&self) -> Result<Option<Link>>;
+
+    /// set the external link for the implementor object
+    fn set_external_link(&mut self, l: Link) -> Result<Option<Link>>;
```
Comment (neithernut) [inline libimaglink/src/external.rs]: The semantics are a bit unclear here. After reading the implementation one might guess that the returned link is the one that was replaced, but a bit of documentation would still be nice.

**Comment by matthiasbeyer** [comment]
As far as I can tell, we don't need the `imag://` scheme, as we do not mix external and internal links... so there is no point in defining this. I will alter the appropriate `doc` section.

**Comment by neithernut** [comment]
We can (and should) still define the `imag://` scheme for use in (user-)interfaces and external programs.

**Comment by matthiasbeyer** [comment]
You mean it should be supported but not be used internally?

**Comment by neithernut** [comment]
Yes.

**Comment by matthiasbeyer** [comment]
I agree on that, though I'm not sure how (and where) to implement this at the current state.

**Comment by neithernut** [comment]
Leaving a specification of the scheme is sufficient for now.

**Pull request merged**